Add initial set of FAQ (#149)
fluxcd/toolkit: docs/faq/index.md

# Frequently asked questions

## General questions

### What does the GitOps Toolkit mean for Flux?

Flux v1 is a monolithic do-it-all operator, the GitOps Toolkit separates the functionalities into specialized controllers.

> Flux v2 will be a curated set of configuration for the GitOps Toolkit which you can simply consume using the tk command. You can easily pick and choose the functionality you need and add extensions to enhance the Toolkit’s functionality. (previous: Limiting the use of features and/or adding extensions on top of Flux has never been this easy.)

Reviewer: ```suggestion
Flux v2 will be a curated set of configuration for the GitOps Toolkit which you can simply consume using the tk command. You can easily pick and choose the functionality you need and add extensions to enhance Flux functionality.
```

> ### Why did you essentially rewrite Flux?

Reviewer: ```suggestion
### Why did you rewrite Flux?
```

> The GitOps Toolkit fully embraces the microservices paradigm: the code will live in very targeted controllers, which allows us to implement new features much more easily.

Reviewer: ```suggestion
Flux was made to reconcile the cluster state with Git, using a single repository as the source of truth for infrastructure and workloads alike. We've seen that Flux users want a more flexible approach, where the cluster state can be composed from different sources taking into account multi-tenancy. 

To accommodate such features we decided to break down Flux into specialised components and rebuild it with modern tooling.
```

> Built with the Kubernetes controller-runtime libraries these specialised controllers can be dynamically configured with Kubernetes custom resources either by cluster admins or by other automated tools.

Reviewer: ```suggestion
The GitOps Toolkit controllers are built with the Kubernetes controller-runtime libraries, these specialised controllers can be dynamically configured with Kubernetes custom resources either by cluster admins or by other automated tools.
```

| RPC endpoint can be configured to a 3rd party solution like FluxCloud to be forwarded as notifications to e.g. Slack | Toolkit components can be configured to POST the events to a `notification-controller` endpoint. Selective forwarding of POSTed events as notifications using `Provider` and `Alert` custom resources.
| Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms
| Unstructured logging | Structured logging for all components
| Custom Prometheus metrics | Generic / common `controller-runtime` Prometheus metrics

### How can I get involved?

There are a variety of ways and we look forward to having you on board building the future of GitOps together with you:

- [Discuss the direction](https://github.com/fluxcd/toolkit/discussions) of the GitOps Toolkit with us
- Join us in #flux-dev on the [CNCF Slack](https://slack.cncf.io)
- Check out our [contributor docs](https://toolkit.fluxcd.io/contributing/)
- Take a look at the [roadmap of the GitOps Toolkit](https://toolkit.fluxcd.io/roadmap/)

### Are there any breaking changes?

> - In Flux v1 Kustomize support was implemented through `.flux.yaml` files in the Git repository. As indicated in the comparison table above, while this approach worked, we found it to be error-prone and hard to debug. The new [Kustomization CR](https://github.com/fluxcd/kustomize-controller/blob/master/docs/spec/v1alpha1/kustomization.md) should be much easier to implement. Unfortunately we needed to drop the support for custom commands.

Reviewer: ```suggestion
- In Flux v1 Kustomize support was implemented through `.flux.yaml` files in the Git repository. As indicated in the comparison table above, while this approach worked, we found it to be error-prone and hard to debug. The new [Kustomization CR](https://github.com/fluxcd/kustomize-controller/blob/master/docs/spec/v1alpha1/kustomization.md) should make troubleshooting much easier. Unfortunately we needed to drop the support for custom commands as running arbitrary shell scripts in-cluster poses serious security concerns.
```

> - Helm users: we redesigned the `HelmRelease` API, so upgrading to using the `helm-controller` will require a little work from you, but you will gain more flexibility, better observability and better performance.

Reviewer: ```suggestion
- Helm users: we redesigned the `HelmRelease` API, so upgrading to `HelmRelease` v2 will require a little work from you, but you will gain more flexibility, better observability and performance.
```

# Frequently asked questions

## General questions

### What does the GitOps Toolkit mean for Flux?

Flux v1 is a monolithic do-it-all operator, the GitOps Toolkit separates the functionalities into specialized controllers.

Flux v2 will be a curated set of configuration for the GitOps Toolkit which you can simply consume using the tk command. You can easily pick and choose the functionality you need and add extensions to enhance Flux functionality.

### Why did you rewrite Flux?

The GitOps Toolkit implements its functionality in individual controllers, which allowed us to implement long-standing feature requests much more easily.

> The GitOps Toolkit controllers are built with the Kubernetes controller-runtime libraries, these specialised controllers can be dynamically configured with Kubernetes custom resources either by cluster admins or by other automated tools.

Author: > > The GitOps Toolkit controllers are built with the Kubernetes controller-runtime libraries
> 
> Why do I (the reader) care about this detail?

Good point. Maybe we could say something about the good things we get for free by basing the controllers on top of this?

This gave us the opportunity to build the GItOps Toolkit with the top Flux feature requests in mind:

- Multi-repository support
- Operational insight through health checks, events and alerts
- Multi-tenancy capabilities

On top of that, testing the GitOps Toolkit and understanding the codebase becomes a lot easier.

> ### What are significant new features of the GitOps Toolkit?

Author: You're right. This section is more about differences and I'll indicate that. Maybe we could add a section on before listing a couple of key features.

---------------------------------- | ----------------------------------
Declarative config through `.flux.yaml` files in the Git repository | Declarative config through a `Kustomization` custom resource, consuming the artifact from the GitRepository
Manifests are generated via shell exec and then reconciled by `fluxd` | Generation, server-side validation, and reconciliation is handled by a specialised `kustomize-controller`
Reconciliation using the service account of the Flux deployment | Support for service account impersonation
Garbage collection needs cluster role binding for Flux to query the Kubernetes discovery API | Garbage collection needs no cluster role binding or access to Kubernetes discovery API
Support for custom commands and generators executed by fluxd in a POSIX shell | No support for custom commands

#### Helm integration

Flux v1                            | Toolkit component driven "Flux v2"
---------------------------------- | ----------------------------------
Declarative config in a single Helm custom resource | Declarative config through `HelmRepository`, `HelmChart` and `HelmRelease` custom resources

> Chart synchronisation embedded in the operator | Extensive release configuration options, and a reconciliation interval per source

Author: Hum... :man_shrugging: I stole this from the upcoming KubeCon presentation slides. It's still in there. :laughing:

Support for fixed SemVer versions from Helm repositories | Support for SemVer ranges for `HelmChart` resources
Git repository synchronisation on a global interval | Planned support for charts from GitRepository sources
Limited observability via the status object of the HelmRelease resource | Better observability via the HelmRelease status object, Kubernetes events, and notifications
Resource heavy, relatively slow | Better performance

#### Notifications, webhooks, observability

Flux v1                            | Toolkit component driven "Flux v2"
---------------------------------- | ----------------------------------
Emits "custom Flux events" to a RPC endpoint | Emits Kubernetes events for all custom resources part of the Toolkit
RPC endpoint can be configured to a 3rd party solution like FluxCloud to be forwarded as notifications to e.g. Slack | Toolkit components can be configured to POST the events to a `notification-controller` endpoint. Selective forwarding of POSTed events as notifications using `Provider` and `Alert` custom resources.

> Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms

Author: Maybe "Well-documented webhook receiver functionality" or some such on the right side instead? https://toolkit.fluxcd.io/guides/webhook-receivers/

> Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms

Reviewer 2: I think the differentiating thing is that they are in the toolkit, rather than off to the side.

> Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms

Reviewer 2: ```suggestion
Webhook receiver is a side-project | Webhook receiver, handling a wide range of platforms, is included
```

> Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms

Reviewer 2: @stefanprodan Are you cool with that statement on webhooks? (flux-recv does actually implement a few formats, including those from some image registries -- the more salient difference is that it's still a bit of a pain to integrate with flux)

- [Discuss the direction](https://github.com/fluxcd/toolkit/discussions) of the GitOps Toolkit with us
- Join us in #flux-dev on the [CNCF Slack](https://slack.cncf.io)
- Check out our [contributor docs](https://toolkit.fluxcd.io/contributing/)
- Take a look at the [roadmap of the GitOps Toolkit](https://toolkit.fluxcd.io/roadmap/)

### Are there any breaking changes?

- In Flux v1 Kustomize support was implemented through `.flux.yaml` files in the Git repository. As indicated in the comparison table above, while this approach worked, we found it to be error-prone and hard to debug. The new [Kustomization CR](https://github.com/fluxcd/kustomize-controller/blob/master/docs/spec/v1alpha1/kustomization.md) should make troubleshooting much easier. Unfortunately we needed to drop the support for custom commands as running arbitrary shell scripts in-cluster poses serious security concerns.
- Helm users: we redesigned the `HelmRelease` API, so upgrading to `HelmRelease` v2 will require a little work from you, but you will gain more flexibility, better observability and performance.

> 

Author: I think I addressed this in the section above. WDYT? The section will change anyway once we have a transition document available.

> 

Reviewer 2: It's a living document as they say, yes true

# Frequently asked questions

## General questions

### What does the GitOps Toolkit mean for Flux?

> Flux v1 is a monolithic do-it-all operator, the GitOps Toolkit separates the functionalities into specialized controllers.

Reviewer 2: Style alert! Either use a conjunction (i.e., "and") or a semicolon to join these clauses together. Or better I think, just have two sentences.

Reviewer 2: ```suggestion
Flux v1 is a monolithic do-it-all operator; the GitOps Toolkit separates the functionalities into specialized controllers.
```

> Flux v2 will be a curated set of configuration for the GitOps Toolkit which you can install and operate simply using the `tk` command. You can easily pick and choose the functionality you need and add extensions to enhance Flux functionality.

Reviewer 2: ```suggestion
Flux v2 will be a curated configuration of the GitOps Toolkit, which you can install and operate simply using the `tk` command. You can easily pick and choose the functionality you need and extend it to serve your own purposes.
```

The timeline we are looking at right now is:

> 1. Put Flux v1 into maintenance mode (no new features being added).

Reviewer 2: ```suggestion
1. Put Flux v1 into maintenance mode (no new features being added; bugfixes and CVEs patched only).
```

1. Continue work on [GitOps Toolkit roadmap](https://toolkit.fluxcd.io/roadmap/).
1. We will provide transition guides for specific user groups, e.g. users of Flux v1 in read-only mode, or of Helm Operator v1, etc. once the functionality is integrated in the GitOps Toolkit and it's deemed "ready".

> 1. Once the use-cases of Flux v1 are covered, we will continue supporting Flux v1 for 6 months. This will be the transition period before it will reach it's end of life.

Reviewer 2: ```suggestion
1. Once the use-cases of Flux v1 are covered, we will continue supporting Flux v1 for 6 months. This will be the transition period before it's considered unsupported.
```

### Why did you rewrite Flux?

> The GitOps Toolkit implements its functionality in individual controllers, which allowed us to implement long-standing feature requests much more easily.

Reviewer 2: ```suggestion
The GitOps Toolkit implements its functionality in individual controllers, which allowed us to address long-standing feature requests much more easily.
```

> By basing these controllers on modern Kubernetes tooling (`controller-runtime` libraries), they can be dynamically configured with Kubernetes custom resources either by cluster admins or by other automated tools and you get increased observability.

Reviewer 2: ```suggestion
By basing these controllers on modern Kubernetes tooling (`controller-runtime` libraries), they can be dynamically configured with Kubernetes custom resources either by cluster admins or by other automated tools -- and you get greatly increased observability.
```

This gave us the opportunity to build the GitOps Toolkit with the top Flux feature requests in mind:

> - Multi-repository support

Reviewer 2: ```suggestion
- Supporting multiple source Git repositories
```

- Operational insight through health checks, events and alerts

> - Multi-tenancy capabilities

Reviewer 2: ```suggestion
- Multi-tenancy capabilities, like applying each source repository with its own set of permissions
```

Limited observability via the status object of the HelmRelease resource | Better observability via the HelmRelease status object, Kubernetes events, and notifications
Resource heavy, relatively slow | Better performance

#### Notifications, webhooks, observability

Flux v1                            | Toolkit component driven "Flux v2"
---------------------------------- | ----------------------------------
Emits "custom Flux events" to a webhook endpoint | Emits Kubernetes events for all custom resources part of the Toolkit
RPC endpoint can be configured to a 3rd party solution like FluxCloud to be forwarded as notifications to e.g. Slack | Toolkit components can be configured to POST the events to a `notification-controller` endpoint. Selective forwarding of POSTed events as notifications using `Provider` and `Alert` custom resources.
Limited incoming webhook functionalities | Extensive support for incoming webhooks for a wide range of platforms
Unstructured logging | Structured logging for all components
Custom Prometheus metrics | Generic / common `controller-runtime` Prometheus metrics

### How can I get involved?

> There are a variety of ways and we look forward to having you on board building the future of GitOps together with you:

Reviewer 2: ```suggestion
There are a variety of ways and we look forward to having you on board building the future of GitOps together:
```
